File: Largest Rectangle in Histogram.java
public class Solution {
    public int largestRectangleArea(int[] arrHeight) {
        Deque<Integer> ascend = new ArrayDeque<Integer>();
        int maxArea = 0;
        List<Integer> height = new ArrayList<Integer>();
        for (int i = 0; i < arrHeight.length; i++) {
            height.add(arrHeight[i]);
        }
        height.add(0);
        for (int i = 0; i < height.size(); i++) {
            if (i > 0 && height.get(i) < height.get(i-1)) {
                while (!ascend.isEmpty()) {
                    int h = height.get(ascend.getLast());
                    if (h < height.get(i))
                        break;
                    ascend.removeLast();
                    int right = i;
                    int left = (ascend.isEmpty()?-1:ascend.peekLast()) + 1;
                    int area = h * (right - left);
                    if (area > maxArea)
                        maxArea = area;
                }
            }
            ascend.add(i);
        }
        return maxArea;
    }
}

File: Largest Rectangle in Histogram.py
class Solution:
	# @param height, a list of integer
	# @return an integer
	def largestRectangleArea(self, height):
		ascend = []
		height.append(0)
		maxArea = 0
		for i in xrange(len(height)):
			if (i > 0 and height[i] < height[i-1]):
				while (len(ascend) > 0):
					h = height[ascend[-1]]
					if (h < height[i]):
						break;
					del ascend[-1]
					right = i
					if (len(ascend) == 0):
						left = 0
					else:
						left = ascend[-1] + 1
					area = h * (right - left)
					if (area > maxArea):
						maxArea = area
			ascend.append(i)
		return maxArea